packages/plugins/src/hardhat/simulate.ts

```diff
@@ -350,17 +350,54 @@ export const simulateDeploymentSubtask = async (
-      // We handle it specifically and immediate error since we want to make sure the user gets an accurate message
-      // indicating they may need to switch to a paid rpc provider for the network where this error occurred.
-      if (e.message.toLowerCase().includes('too many requests')) {
+      /**
+       * There are really only a few cases where an error will be throw during the deployment and be caught
+       * here (that we know of):
+       * 1. There's a legitimate error in our execution logic
+       * 2. We estimate the merkle leaf gas incorrectly
+       * 3. We fail to find a valid batch size
+       * 4. The users RPC provider rate limits us
+       *
+       * Retrying execution won't help in cases 1, 2, or 3.
+       *
+       * We have this retry logic implemented almost entirely for case 4 where we hit a rate limit that
+       * causes the execution to fail. We found that simply retrying in this case generally does work, but
+       * is not 100% reliable. We often get other errors that appear to be caused by the initial rate limiting.
+       *
+       * For example, we occassionally encountered a situation where transactions would fail after rate
+       * limiting due to the nonce used in the transaction being incorrect. There's also another case where
+       * the transaction to deploy the users Safe would fail after rate limiting because the Safe has already
+       * been deployed. This is something we have logic specifically to prevent that we know works well.
+       *
+       * We believe these issues are related to there being data cached somewhere that is not correct. We
+       * found the most reliable method to resolve these sort of issues was to completely reset the simulation
+       * fork back to the original state using `hardhat_reset`.
+       */
+
+      if (!hre.config.networks.hardhat.forking) {
         throw new Error(
-          `Simulation failed due to rate limiting on RPC provider for ${fetchNameForNetwork(
-            BigInt(deployment.chainId)
-          )}. You may need to switch to an RPC provider that supports higher request volume for this network.`
+          'Simulation was not using fork. This is a bug, please report it to the developers.'
         )
       }
 
+      await provider.send('hardhat_reset', [
+        {
+          forking: {
+            jsonRpcUrl: hre.config.networks.hardhat.forking?.url,
+            blockNumber: hre.config.networks.hardhat.forking?.blockNumber,
+          },
+        },
+      ])
+
+      // Since we're resetting back to the initial state, we also have to fund the simulation sender wallet
+      if (
+        executionMode === ExecutionMode.LocalNetworkCLI ||
+        executionMode === ExecutionMode.Platform
+      ) {
+        signer = new ethers.Wallet(getSphinxWalletPrivateKey(0), provider)
+        await fundAccountMaxBalance(signer.address, provider)
+      }
```

Comment:
Makes sense, but I'm probably going to forget to update this code if we add more pre-simulation steps in the future. Would be less of a maintenance hazard if this code and the [corresponding code](https://github.com/sphinx-labs/sphinx/blob/265060df7c542cb4d02973e9fa77c4778e57f16e/packages/plugins/src/hardhat/simulate.ts#L247-L248) in `simulateDeploymentSubtask` are refactored into a single function